### ISSUE TITLE
xsl:context-item Code Coverage Non-Conformance

### ISSUE THREAD

**Issue by birdya22** [xspec/xspec#1938]
### Current Non-Conformance
The `xslt-code-coverage-behavior-by-element.md` file says that `xsl:context-item` elements should always be marked as 'ignored'.
The `coverage-report.xsl` in 3.0.3 marks them as 'hit' due to a specific rule:
```
<xsl:when test="self::xsl:context-item">
  <!-- Saxon does not seem to call enter() for xsl:context-item (xspec/xspec#1410) -->
  <xsl:sequence select="local:coverage(parent::xsl:template, $module-id)" />
</xsl:when>
```
Issue [1410](https://github.com/xspec/xspec/issues/1410) was specifically about xsl:context-item not being marked as 'hit'. This issue will change the result of 1410.

### Changes in `coverage-report.xsl`
The current rule needs to be reviewed in `coverage-report.xsl` to ensure `xsl:context-item` elements are always marked as 'ignored'.

### Changes in `test/end-to-end/cases-coverage/expected`
In the coverage report `xsl-context-item-01-coverage.html` the `xsl:context-item` element will be marked as 'ignored' rather than 'hit'.
In the coverage report external_issue-1410_invoked-empty-coverage.html both `xsl:context-item` elements will be marked as 'ignored'. The first is currently 'hit' and the second 'missed'.

**Comment by galtm** [comment]
This change should be straightforward to make. I'm looking for opinions on two things:

1. @xspec/xspec : Do you like this design change? I don't think I have a strong opinion either way.
2. Would it need to be in a version where we increment the first part of the x.y.z version number? On one hand, marking `xsl:context-item` as  'ignored' instead of using the 'hit' or 'missed' status of the parent element would be a change in the informational content of the report (and not a bug fix). On the other hand, I don't really consider the HTML report of code coverage to be a public API, and there will be no change in users' ability to generate such reports for their code. I am inclined to say that this kind of change would be acceptable in XSpec 3.1 and not have to wait for XSpec 4, but other input would be helpful.

**Comment by birdya22** [comment]
I'm assuming @xspec/xspec was meant to be me.

> 1. @xspec/xspec : Do you like this design change? I don't think I have a strong opinion either way.

It is slightly harder to understand given that you need to know where the tree navigation is happening to know when the `category-based-on-trace-data accumulator` changes.
As a more general comment I did struggle to understand how it worked when I first had an issue and looked at it. I struggled with the fact that it parsed the stylesheet using the regex and iterated around the result whilst also navigating in sync through the stylesheet as an xml document, and looking up hits in the `coverage.xml` file based on the nodes in the xml document. Without that initial understanding I think people would struggle to produce their own coverage-report based on this the XSpec one.
What you've done on replicating the Rules in their own templates is good, but I'm not sure the use of the `category-based-on-trace-data` accumulators adds much as it still relies on calling the `local:hits-on-node` function.

> 2. Would it need to be in a version where we increment the first part of the x.y.z version number?

I'm inclined to agree about including it in XSpec 3.1. 
Also, assuming that XSpec 3.1 comes out after Saxon 12.5 (which I assume is very very likely) then 3.1 could be described as code coverage updates based on changes in recent versions of Saxon, some of which have changed the way code coverage reports look.
Are you intending many other changers in XSpec 3.1?

**However**, based on its [description](https://www.saxonica.com/documentation12/index.html#!xsl-elements/context-item)  "Used to declare the initial context item for a template: whether the template requires a context item, and if so, what its expected type is." it is a Declaration rather than an Instruction even though the page doesn't give it a Category. On that basis I said "Always Ignore". 
However it is slightly different in that it isn't a direct child of the `xsl:stylesheet` or `xsl:transform` so what do you think about leaving it as it is in coverage-report.xsl and changing the rule to Use Parent Data? 
It is the only element with a "Always Ignore" Rule that isn't a child of the `xsl:stylesheet` or `xsl:transform`, or a child of an "Always Ignore" rule element.
On that basis it could be a special case of a Declaration that isn't "Always Ignore".

**Comment by galtm** [comment]
> I'm assuming @xspec/xspec was meant to be me.
> 
> > 1. @xspec/xspec : Do you like this design change? I don't think I have a strong opinion either way.

@birdya22 : No, actually, @xspec/xspec is a way to reach the "team" for this repository. I wanted to ask others who have been involved with this repository if they have input on the design change regarding `xsl:context-item` (namely, always ignore instead of hit/miss) that this issue is about. I meant to address question 2 both to that list of contributors and to you. I didn't intend to ask you question 1 because the answer seemed obvious; surely you like a design that you're proposing. :)

**Comment by galtm** [comment]
> Are you intending many other changers in XSpec 3.1?

I have in the back of my mind that it would be good to improve the accessibility of the HTML report of test results, but I haven't had time to delve into it yet, so I don't know if it will be in 3.1. Everyone who contributes to this repo is doing so in their spare time, which makes release planning challenging...

**Comment by galtm** [comment]
> it is a Declaration rather than an Instruction

> However it is slightly different in that it isn't a direct child of the `xsl:stylesheet` or `xsl:transform`

Yes. Those two points lead in different directions, and that's why I'm currently flip-flopping, without a strong opinion.

I have a slight leaning toward the Use Parent Data rule because visually, an "ignored" line surrounded by "hit" or "miss" lines draws attention in a way that might not be so helpful. For instance, in the following mock-up, line 15 might make me think, "Why is that one line different from what's around it?" which is a distraction from "That template in red either isn't tested or is unneeded code that I should delete."

![image](https://github.com/xspec/xspec/assets/40716346/08d6e560-ed08-445e-85b7-2814ac25773a)

**Comment by birdya22** [comment]
I agree.
What I wrote about "Ignored" was "Ignored - the node is a declaration and no trace information is provided. Without extensive work in XSpec to parse the stylesheet it is not possible to determine if these declarations are used".
It is possible to mark it as 'hit' or 'missed' with the Use Parent Rule without extensive work (although I didn't write it down I was thinking of cases like this when I wrote "extensive work": are the functions `accumulator-before` or `accumulator-after` used for a specific `xsl:accumulator`. If so mark it as a 'hit', otherwise mark it as 'missed').
